File: python/gilded_rose.py
# -*- coding: utf-8 -*-

from abc import ABC, abstractmethod

class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)


class ItemUpdater(ABC):
    def __init__(self, item):
        self.item = item
    
    def update(self):
        self.item.sell_in -= 1
        self.item.quality += self.quality_delta()
        if self.item.sell_in < 0:
            self.item.quality += self.expired_delta()
        self.item.quality = max(0, min(50, self.item.quality))
    
    @abstractmethod
    def quality_delta(self):
        pass
    
    def expired_delta(self):
        return 0


class NormalItemUpdater(ItemUpdater):
    def quality_delta(self):
        return -1
    
    def expired_delta(self):
        return -1


class AgedBrieUpdater(ItemUpdater):
    def quality_delta(self):
        return 1
    
    def expired_delta(self):
        return 1


class SulfurasUpdater(ItemUpdater):
    def quality_delta(self):
        return 0
    
    def update(self):
        self.item.quality = 80


class BackstagePassUpdater(ItemUpdater):
    def quality_delta(self):
        if self.item.sell_in > 10:
            return 1
        return 2 if self.item.sell_in > 5 else 3
    
    def expired_delta(self):
        return -self.item.quality


class GildedRose:
    UPDATERS = {
        "Aged Brie": AgedBrieUpdater,
        "Sulfuras, Hand of Ragnaros": SulfurasUpdater,
        "Backstage passes to a TAFKAL80ETC concert": BackstagePassUpdater
    }
    
    def __init__(self, items):
        self.items = items
    
    def update_quality(self):
        for item in self.items:
            self.UPDATERS.get(item.name, NormalItemUpdater)(item).update()

File: js-jasmine/src/gilded_rose.js
class Item {
  constructor(name, sellIn, quality) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

class Shop {
  constructor(items = []) {
    this.items = items;
  }

  updateQuality() {
    for (var i = 0; i < this.items.length; i++) {

      const item = this.items[i];

      if (item.name.startsWith("Conjured")) {
        updateConjuredItem(item);
      }
      else if (item.name.startsWith("Sulfuras")) {
        updateSulfuras(item);
      }
      else {
          switch (item.name) {
            case "Aged Brie":
              updateAgedBrie(item);
              break;

            case "Backstage passes to a TAFKAL80ETC concert":
              updateBackstagePass(item);
              break;

            default:
              updateNormalItem(item);
              break;
          }
      }
    }

    return this.items;
  }
}

function updateNormalItem(item) {
  item.sellIn = item.sellIn - 1;

  if (item.sellIn >= 0) {
    item.quality = item.quality - 1;
  } 
    else {
      item.quality = item.quality - 2;
    }

  if (item.quality < 0) {
    item.quality = 0;
  }
}

function updateAgedBrie(item) {
  item.sellIn = item.sellIn - 1;

  if (item.sellIn >= 0) {
    item.quality = item.quality + 1;
  } 
    else {
      item.quality = item.quality + 2;
    }

  if (item.quality > 50){
    item.quality = 50;
  } 
}

function updateBackstagePass(item) {
  item.sellIn = item.sellIn - 1;

  if (item.sellIn < 0) {
    item.quality = 0;
  } 
    else {
      item.quality = item.quality + 1;

      if (item.sellIn <= 10) {
        item.quality = item.quality + 1;
      }
      if (item.sellIn <= 5) {
        item.quality = item.quality + 1;
      }
    }

  if (item.quality > 50){
      item.quality = 50;
  } 
}

function updateSulfuras(item) {
  item.quality = 80;
}

function updateConjuredItem(item) {
  item.sellIn = item.sellIn - 1;

  if (item.sellIn >= 0) {
    item.quality = item.quality - 2;
  } 
    else {
      item.quality = item.quality - 4;
    }

  if (item.quality < 0) item.quality = 0;
}

module.exports = {
  Item,
  Shop,
};

